### PR TITLE
Multiple HLS codec and bitrate fixes (10.8.z)

### PR DESCRIPTION
**Pull request by Shadowghost**
Backports #9388 and #8546

**Changes**
* Enforce HLS audio and video codec restrictions on transcoding profiles
* Properly calculate bitrates for transcoded codecs
* Fix encoder check for DTS
* Add encoder check for TrueHD
* Add decoder check for DTS and TrueHD
* Fix condition check in CanStreamCopyAudio
* Add FLAC/Opus workaround for HLS

### CODE REVIEW

File: MediaBrowser.Model/Dlna/StreamBuilder.cs
Comment (nyanmisaka) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:827]: We currently do not support vorbis in HLS.

https://github.com/jellyfin/jellyfin/blob/edc627fd5b5a5bd19c843dd9e2970b1ebce3fbfd/Jellyfin.Api/Helpers/DynamicHlsHelper.cs#L598-L637

**Review by nyanmisaka — APPROVED** [review]
Tested in both TS and fMP4.

**Review by crobibero — APPROVED** [review]

**Comment by joshuaboniface** [comment]
So curiously, I've had a user test this and it seems like this alone isn't enough to fix the issue they were seeing. I'm planning to do some more extensive testing over the coming days to see how much I can replicate and confirm. Hold off on merge for a bit longer.

**Comment by holow29** [comment]
Where are you getting supported audio/video codecs? Looking at RFC 8216, I see some information, but can't find reference to FLAC, for example, even though I know it works in HLS in fmp4. I'm mostly wondering because I'm not sure I've come across many codecs that will work in mp4 but won't in HLS fmp4, though more testing is probably merited.

For example, while devising profile for Swiftfin (https://github.com/jellyfin/Swiftfin/pull/519), I included a lot of codecs for HLS fmp4 that VLCKit supports to avoid transcoding for cases like TrueHD audio. With these server-side HLS restrictions, there will be cases like old Blurays with VC-1 or mpeg2video and TrueHD audio that will force a transcode instead of DirectStream. Obviously, I should do more testing to make sure these codecs will actually play in HLS fmp4, and there is an argument to be made that Jellyfin should be adhering to the specs instead of what is possible or non-standard.

**Comment by nyanmisaka** [comment]
> 2.25. The container format for xHE-AAC, Apple Lossless, and FLAC audio MUST be fMP4.

Since HLS is an Apple standard, you should always refer to [their docs](https://developer.apple.com/documentation/http_live_streaming/http_live_streaming_hls_authoring_specification_for_apple_devices).

**Comment by holow29** [comment]
Okay, I don't see opus mentioned in their docs, but it is in this PR. Is your position that Jellyfin server should restrict HLS based on Apple's authoring specs (vs. the standard defined in RFC 8216 or what individual clients can handle/request in their transcoding profile) ?

**Comment by nyanmisaka** [comment]
Shaka-player. https://github.com/jellyfin/jellyfin-web/pull/4041

It can handle almost all the video and audio codecs supported by your browser. OPUS is no exception too.

**Comment by nyanmisaka** [comment]
Use Apple specs as baseline and extend more codecs by the clients we used.
We may use AV1 in HLS in the future, which is certainly not supported by Apple for the time being.

**Comment by holow29** [comment]
It really seems to me like relying on client device profile is a better solution; maybe there is a reason that won't work here that I'm not thinking of. However, I think the server should only be limiting what it can and can't handle given the restriction can actually be placed client-side. The server (and ffmpeg) can handle passing many other formats in HLS and fmp4. Shaka player, VLCkit-based players, etc. can handle things like opus and others that are not defined in Apple's HLS documentation/appendices. Is the idea to only support codecs that all clients can use?

In my example above for Swiftfin, if this PR merges, there will be no way to play old Blurays with VC-1 video or mpeg2video and TrueHD audio without transcoding the video - even though that isn't a limitation of the player itself. (Again, I need to test again because I don't recall, but I am pretty sure I was able to pass VC-1 and mpeg2video in a fmp4 HLS container). I also need to test DTS in fmp4 - which I think I've gotten to work before - because if this merges and Shadowghost's BD/DVD fix merges then Blurays will need to be streamed, and using HLS, any Bluray with DTS audio will automatically be forced to transcode it. Basically, I think this largely limits what will be able to remux over HLS, forcing either DirectStream or transcoding, when it is not a client limitation.

**Comment by nyanmisaka** [comment]
@Shadowghost Better to add a `IsCopyCodec()` check before applying these restrictions.

We only need to ensure that the server can re-encode these codecs. Let the client handle the remuxing cases.

**Comment by Shadowghost** [comment]
@holow29 the reason we are introducing this is that we had an issue where a client accidentially sent a codec profile not compliant with the HLS spec (DTS via HLS-TS) leading to no audio. This is reproducible if you try to send DTS via HLS (VLC on windows can't play those files, MPV can).
Most usecases should be fixed by addind a copycodec flag though since the main problem is if the sorce (audio or video) gets transcoded.

**Comment by joshuaboniface** [comment]
OK definitely seems like we need the whole branch, I was able to replicate the issue still with just this PR. So far I haven't seen it introduce any new issues but you never know. But I'm willing to take the risk since this is a pretty major breakage.

**Comment by Shadowghost** [comment]
I'll rebase the whole master PR (minus the VBR stuff) once we sorted out the remaining things.

**Review by Bond-009 — APPROVED** [review]

File: Jellyfin.Api/Helpers/DynamicHlsHelper.cs
Comment (crobibero) [inline Jellyfin.Api/Helpers/DynamicHlsHelper.cs:758]: This seems to expect that `flac` is always the last audio codec appended
I'm no expert on the playlist code but that seems like a weird limitation to have?

It's also very possible that I am totally not understanding the contents of `srcPlaylist`

Comment (Shadowghost) [reply Jellyfin.Api/Helpers/DynamicHlsHelper.cs:758]: We're directly modifying the HLS playlist which only includes one audio codec, so this is works as intended.

Comment (crobibero) [reply Jellyfin.Api/Helpers/DynamicHlsHelper.cs:758]: For future readers this is what is being modified

```
#EXT-X-STREAM-INF:BANDWIDTH=8000000,VIDEO-RANGE=SDR,CODECS="avc1.640033,flac",RESOLUTION=1920x1080,FRAME-RATE=59.94
```

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Model/Dlna/StreamBuilder.cs
Comment (Shadowghost) [reply MediaBrowser.Model/Dlna/StreamBuilder.cs:28]: If they work on fmp4 we can do that?

**Review by nyanmisaka — APPROVED** [review]

**Comment by holow29** [comment]
Was the decision made to not use `IsCopyCodec()` check then? I think this will affect allowing codecs like vc1 and mpeg2video being passed untouched in HLS.

**Comment by Shadowghost** [comment]
`IsCopyCodec()` can't be used in the places where the check would be needed, this was the best solution I can find. Officially HLS only supports h264 and h265 as video codec, if we need to add more to the list, we can do it in another PR.
Your thoughts on this @nyanmisaka?

**Comment by nyanmisaka** [comment]
@holow29 Feel free to open a PR to add them in these if you support them on the client side.
- `EncodingHelper.cs/shiftVideoCodecs`
- `StreamBuilder.cs/_supportedHlsVideoCodecs`

**Pull request merged**
